File: Array/zhangjunwei/arrayNesting_6_22.c
//
// Created by 张俊伟 on 2020/6/22.
//
#include <stdbool.h>
#include <stdlib.h>
#include <math.h>
#include <assert.h>
#include <stdio.h>
int arrayNesting(int *nums, int numsSize) {
    bool *visted = (bool *) calloc(numsSize, sizeof(bool));
    assert(visted!=NULL);
    int maxSize = 0;
    for (int i = 0; i < numsSize; ++i) {
        if (!*(visted + i)) {
            int count =0, start = *(nums + i);
            do {
                *(visted + start) = true;
                count++;
                start = *(nums + start);
            } while (start != *(nums + i));
            if (maxSize < count)
                maxSize = count;
        }
    }
    free(visted);
    return maxSize;
}
int main(){
    int nums[] = {5,4,0,3,1,6,2};
  int res = arrayNesting(nums, 7);
  printf("res: %d ",res);
}

File: Array/zhangjunwei/arrayNesting_6_22.java
package com.lanfairy.md;

import android.util.ArraySet;
import android.util.SparseIntArray;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

public class Solution {
    public int arrayNesting1(int[] nums) {
        int size = nums.length;
        Set<Integer> set = new TreeSet<>();
        int sMaxNum = 0;//最大的集合有的元素个数
        for (int i = 0; i < size; i++) {
            int j = i;

            while (!set.contains(nums[j])) {
                set.add(nums[j]);
                j = nums[j];
            }
            if (sMaxNum < set.size())
                sMaxNum = set.size();
            set.clear();
        }

        return sMaxNum;

    }

    public int arrayNesting(int[] nums) {
        int size = nums.length;
        int sMaxNum = 0;//最大的集合有的元素个数
        boolean[] visted = new boolean[size];
        for (int i = 0; i < size; i++) {
            if (!visted[i]) {
                int start = nums[i];
                int count = 0;
                do {
                    visted[start] = true;
                    start = nums[start];
                    count++;
                }while (start!=nums[i]);

                sMaxNum = Math.max(sMaxNum, count);
            }
        }

        return sMaxNum;

    }

    public static void main(String[] args) {
        int[] num = new int[]{5, 4, 0, 3, 1, 6, 2};

        int res = new Solution().arrayNesting(num);
        System.out.println(res);
    }

}
